magefile.go

```diff
@@ -1295,3 +1321,123 @@ func shouldBuildAgent() bool {
+	const cliName = "gcloud"
+	cmd := exec.CommandContext(ctx, "which", cliName)
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("%s CLI is not installed: %w", cliName, err)
+	}
+
+	// Check if user is already authenticated
+	for authSuccess := false; !authSuccess; {
+		cmd = exec.CommandContext(ctx, cliName, "auth", "list", "--filter=status:ACTIVE", "--format=json")
+		output, err := cmd.Output()
+		if err != nil {
+			return fmt.Errorf("unable to list authenticated accounts: %w", err)
+		}
+
+		var authList []struct {
+			Account string `json:"account"`
+		}
+		if err := json.Unmarshal(output, &authList); err != nil {
+			return fmt.Errorf("unable to parse authenticated accounts: %w", err)
+		}
+
+		if len(authList) > 0 {
+			// We have at least one authenticated, active account. All set!
+			authSuccess = true
+			continue
+		}
+
+		fmt.Fprintln(os.Stderr, "❌  GCP authentication unsuccessful. Retrying...")
+
+		// Try to authenticate user
+		cmd = exec.CommandContext(ctx, cliName, "auth", "login")
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("unable to authenticate user: %w", cliName, err)
+		}
+	}
+
+	return nil
+}
+
+func authESS(ctx context.Context) error {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return fmt.Errorf("unable to determine user's home directory: %w", err)
+	}
+	essAPIKeyFile := filepath.Join(homeDir, ".config", "ess", "api_key.txt")
+
+	_, err = os.Stat(essAPIKeyFile)
+	if os.IsNotExist(err) {
+		if err := os.MkdirAll(filepath.Dir(essAPIKeyFile), 0700); err != nil {
+			return fmt.Errorf("unable to create ESS config directory: %w", err)
+		}
+
+		if err := os.WriteFile(essAPIKeyFile, nil, 0600); err != nil {
+			return fmt.Errorf("unable to initialize ESS API key file: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("unable to check if ESS config directory exists: %w", err)
+	}
+
+	// Read API key from file
+	data, err := os.ReadFile(essAPIKeyFile)
+	if err != nil {
+		return fmt.Errorf("unable to read ESS API key: %w", err)
+	}
+
+	data = bytes.TrimSpace(data)
+	essAPIKey := string(data)
+
+	// Attempt to use API key to check if it's valid
+	for authSuccess := false; !authSuccess; {
+		client := ess.NewClient(ess.Config{ApiKey: essAPIKey})
+		u, err := client.GetUser(ctx, ess.GetUserRequest{})
+		if err != nil {
+			return fmt.Errorf("unable to successfully connect to ESS API: %w", err)
+		}
+
+		if u.User.UserID != 0 {
+			// We have a user. All set!
+			authSuccess = true
+			continue
+		}
+
+		fmt.Fprintln(os.Stderr, "❌  ESS authentication unsuccessful. Retrying...")
+
+		essAPIKey, err = stringPrompt("Please provide a ESS (QA) API key:")
+		if err != nil {
+			return fmt.Errorf("unable to read ESS API key from prompt: %w", err)
+		}
+	}
+
+	// Write API key to file for future use
+	if err := os.WriteFile(essAPIKeyFile, []byte(essAPIKey), 0600); err != nil {
+		return fmt.Errorf("unable to persiste ESS API key for future use: %w", err)
+	}
```

Comment:
I think we should validate that the provided key actually works at this point.

Reply:
I'm not sure I understand.  We're doing that as the very first thing in the loop.  I'll clarify the comment a bit more in the success conditional inside the loop.

Reply:
Done in eb345a605fcd5d0e01bb129efd0f9869ce9d5cf2.